src/main/java/io/cdap/plugin/gcp/gcs/source/GCSSource.java

```diff
@@ -78,6 +77,31 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
     super.configurePipeline(pipelineConfigurer);
   }
 
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    // Get location of the source for lineage
+    String location;
+    Credentials credentials = config.connection.getServiceAccount() == null ?
+      null : GCPUtils.loadServiceAccountCredentials(config.connection.getServiceAccount(),
+                                                    config.connection.isServiceAccountFilePath());
+    Storage storage = GCPUtils.getStorage(config.connection.getProject(), credentials);
+    try {
+      location = storage.get(config.getBucketName()).getLocation();
+    } catch (StorageException e) {
+      throw new RuntimeException(
+        String.format("Unable to access bucket %s. ", config.getBucketName())
+          + "Ensure you entered the correct bucket path and have permissions for it.", e);
+    }
+
+    // create asset for lineage
+    String referenceName = Strings.isNullOrEmpty(config.getReferenceName())
+      ? ReferenceNames.normalizeFqn(config.getPath())
+      : config.getReferenceName();
+    asset = Asset.builder(referenceName)
+      .setFqn(config.getPath()).setLocation(location).build();
+    super.prepareRun(context);
```

Comment:
Should the `super.prepareRun()` method be called first?

Reply:
Don't think so. I wonder if it's the same case for S3 and FTP as well. (need to test them). if I call it first, it'll throw a `NullPointerException` at `io.cdap.plugin.common.LineageRecorder.createExternalDataset(LineageRecorder.java:68)`

Reply:
Can you link to the line that causes the NPE? I want to understand that causes the NPE.

Reply:
I see, thanks for the clarification.